zhaquirks/tuya/ts0601_keypad.py

```diff
@@ -0,0 +1,266 @@
+    ARM_DELAY_TIME_DP_ID = 103  # Integer, 0-180, read/write
+    KEYPAD_BEEPS_DP_ID = 104  # Boolean, read/write
+    QUICK_SOS_DP_ID = 105  # Boolean, read/write
+    QUICK_DISARM_DP_ID = 106  # Boolean, read/write
+    QUICK_ARM_DP_ID = 107  # Boolean, read/write
+    ADMIN_CODE_DP_ID = 108  # String, read only
+    USER_CODE_DP_ID = 109  # String, read only
+    RESET_DP_ID = 110  # Boolean, write only
+    ARM_DELAY_BEEPS_DP_ID = 111  # Boolean, read/write
+
+    attributes = TuyaMCUCluster.attributes.copy()
+    attributes.update(
+        {
+            TUYA_CLUSTER_ID + ARM_DELAY_TIME_DP_ID: ("arm_delay_time", t.uint8_t),
+            TUYA_CLUSTER_ID + ARM_DELAY_BEEPS_DP_ID: ("arm_delay_beeps", t.uint8_t),
+            TUYA_CLUSTER_ID + KEYPAD_BEEPS_DP_ID: ("keypad_beeps", t.uint8_t),
+            TUYA_CLUSTER_ID + QUICK_DISARM_DP_ID: ("quick_disarm", t.uint8_t),
+            TUYA_CLUSTER_ID + QUICK_ARM_DP_ID: ("quick_arm", t.uint8_t),
+            TUYA_CLUSTER_ID + QUICK_SOS_DP_ID: ("quick_sos", t.uint8_t),
+            TUYA_CLUSTER_ID + ADMIN_CODE_DP_ID: ("admin_code", t.CharacterString),
+            TUYA_CLUSTER_ID + USER_CODE_DP_ID: ("user_code", t.CharacterString),
+        }
+    )
+
+    dp_to_attribute: Dict[int, DPToAttributeMapping] = {
+        BATTERY_PERCENTAGE_DP_ID: DPToAttributeMapping(
+            PowerConfigurationCluster.ep_attribute,
+            "battery_percentage_remaining",
+            TuyaDPType.VALUE,
+            converter=lambda x: x * 2,
+        ),
+        ARM_DELAY_TIME_DP_ID: DPToAttributeMapping(
+            TuyaMCUCluster.ep_attribute,
+            "arm_delay_time",
+            TuyaDPType.VALUE,
+        ),
+        ARM_DELAY_BEEPS_DP_ID: DPToAttributeMapping(
+            TuyaMCUCluster.ep_attribute,
+            "arm_delay_beeps",
+            TuyaDPType.BOOL,
+        ),
+        KEYPAD_BEEPS_DP_ID: DPToAttributeMapping(
+            TuyaMCUCluster.ep_attribute,
+            "keypad_beeps",
+            TuyaDPType.BOOL,
+        ),
+        QUICK_DISARM_DP_ID: DPToAttributeMapping(
+            TuyaMCUCluster.ep_attribute,
+            "quick_disarm",
+            TuyaDPType.BOOL,
+        ),
+        QUICK_ARM_DP_ID: DPToAttributeMapping(
+            TuyaMCUCluster.ep_attribute,
+            "quick_arm",
+            TuyaDPType.BOOL,
+        ),
+        QUICK_SOS_DP_ID: DPToAttributeMapping(
+            TuyaMCUCluster.ep_attribute,
+            "quick_sos",
+            TuyaDPType.BOOL,
+        ),
+        ADMIN_CODE_DP_ID: DPToAttributeMapping(
+            TuyaMCUCluster.ep_attribute,
+            "admin_code",
+            TuyaDPType.STRING,
+        ),
+        USER_CODE_DP_ID: DPToAttributeMapping(
+            TuyaMCUCluster.ep_attribute,
+            "user_code",
+            TuyaDPType.STRING,
+        ),
+    }
+
+    def _dp_2_event(self, datapoint: TuyaDatapointData) -> None:
+        """Convert DP to Event."""
+        zone_id = 0
+        user_code = self._attr_cache.get(
```

Comment:
So far, I try to get the user code from the cache, but this assumes that the device has sent the user code - if not, I default to "1234". This might cause issues in some scenarios, but I'm not quite sure how to solve this.